// @neo4j/cypher-builder/dist/pattern/Pattern.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pattern = void 0;
const WithWhere_1 = require("../clauses/mixins/sub-clauses/WithWhere");
const mixin_1 = require("../clauses/utils/mixin");
const label_expressions_1 = require("../expressions/labels/label-expressions");
const RelationshipRef_1 = require("../references/RelationshipRef");
const add_label_token_1 = require("../utils/add-label-token");
const compile_cypher_if_exists_1 = require("../utils/compile-cypher-if-exists");
const escape_1 = require("../utils/escape");
const PartialPattern_1 = require("./PartialPattern");
const PatternElement_1 = require("./PatternElement");
/** Represents a pattern of a single node or n-relationships to be used in clauses.
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/syntax/patterns/)
 * @group Patterns
 */
let Pattern = class Pattern extends PatternElement_1.PatternElement {
    constructor(node, previous) {
        super(node);
        this.withLabels = true;
        this.withVariable = true;
        this.previous = previous;
    }
    withoutLabels() {
        this.withLabels = false;
        return this;
    }
    withoutVariable() {
        this.withVariable = false;
        return this;
    }
    withProperties(properties) {
        this.properties = properties;
        return this;
    }
    related(rel) {
        if (!rel)
            rel = new RelationshipRef_1.RelationshipRef();
        return new PartialPattern_1.PartialPattern(rel, this);
    }
    getVariables() {
        const prevVars = this.previous?.getVariables() ?? [];
        prevVars.push(this.element);
        return prevVars;
    }
    /**
     * @internal
     */
    getCypher(env) {
        const prevStr = this.previous?.getCypher(env) ?? "";
        const nodeRefId = this.withVariable ? `${this.element.getCypher(env)}` : "";
        const propertiesStr = this.properties ? this.serializeParameters(this.properties, env) : "";
        const nodeLabelStr = this.withLabels ? this.getNodeLabelsString(this.element, env) : "";
        const whereStr = (0, compile_cypher_if_exists_1.compileCypherIfExists)(this.whereSubClause, env, { prefix: " " });
        return `${prevStr}(${nodeRefId}${nodeLabelStr}${propertiesStr}${whereStr})`;
    }
    getNodeLabelsString(node, env) {
        const labels = node.labels;
        if (labels instanceof label_expressions_1.LabelExpr) {
            const labelsStr = labels.getCypher(env);
            if (!labelsStr) {
                return "";
            }
            return (0, add_label_token_1.addLabelToken)(env.config.labelOperator, labels.getCypher(env));
        }
        else {
            const escapedLabels = labels.map(escape_1.escapeLabel);
            if (escapedLabels.length === 0) {
                return "";
            }
            return (0, add_label_token_1.addLabelToken)(env.config.labelOperator, ...escapedLabels);
        }
    }
};
exports.Pattern = Pattern;
exports.Pattern = Pattern = __decorate([
    (0, mixin_1.mixin)(WithWhere_1.WithWhere)
], Pattern);
